### ISSUE TITLE
Assigning boolean in Select statement throws invalid cast.

### ISSUE THREAD

**Issue by MikeWasson** [aspnet/EntityFramework#2246]
I have a DbContext like this:

    public class BookContext : DbContext
    {
        public DbSet<Author> Authors { get; set; }
     }

This code throws an invalid cast

    var x = BookContext.Authors
        .Select(x => new
        {
            Text = x.FirstMidName,
            Value = x.AuthorID.ToString(),
            Selected = true  // Throws invalid cast
        }).ToList();

Works fine without 'Selected = true'. 

Stack trace:

```
System.InvalidCastException was unhandled by user code
  HResult=-2147467262
  Message=Specified cast is not valid.
  Source=EntityFramework.Relational
  StackTrace:
       at Microsoft.Data.Entity.Relational.RelationalObjectArrayValueReader.ReadValue[T](Int32 index)
       at lambda_method(Closure , QuerySourceScope )
       at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
       at Microsoft.Data.Entity.Query.LinqOperatorProvider.ExceptionInterceptor`1.EnumeratorExceptionInterceptor.MoveNext()
       at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
       at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
       at ContosoBooks.Controllers.BookController.GetAuthorsListItems(Int32 selected) in C:\Users\mwasson\Documents\Visual Studio 2015\Projects\sample\src\ContosoBooks\Controllers\BookController.cs:line 143
       at ContosoBooks.Controllers.BookController.Create() in C:\Users\mwasson\Documents\Visual Studio 2015\Projects\sample\src\ContosoBooks\Controllers\BookController.cs:line 45
  InnerException: 
```

**Comment by maumar** [comment]
Problem here is that we are sending the following query to the database:

```
SELECT [x].[FirstMidName], [x].[AuthorID], 1
FROM [Author] AS [x]
```

This then gets stored in the value buffer as {string, int, int}, rather than {string, int, bool} - hence the conversion fails, when we try to cast from int to bool.

Possible solution is to evaluate the constant on the client, rather than sending it to server. We need to make sure that we still translate constants that are bound to the source query somehow (part of a more complex expression involving a column, or an argument to a function)

**Comment by divega** [comment]
Wouldn't 'CAST (1 AS BIT)" work?

**Comment by maumar** [comment]
It would, and would be a simpler solution. However the other solution seems to be consistent with our funcletization strategy (funcletize as much as we can if its deterministic and doesn't bring more rows from the database).

**Comment by anpete** [comment]
Yes, better to do client eval here as it means less data on the wire.

**Comment by divega** [comment]
Agreed, but any idea of how much work it would take each way? We need to be careful not to turn this little thing into something big :smile:

**Comment by anpete** [comment]
Well, it shouldn't be too bad as we can already do the client eval. I will follow up with @maumar later today for an update.

**Comment by maumar** [comment]
I think we got a very easy/local fix for this

**Comment by maumar** [comment]
Fixed in e8e50c0de94dd72ff9785e94f474466c2cfa5cad
